### PR TITLE
vmap: better errors for mismatched axis in keyword arguments

### LINKED ISSUE #8883: hit assertion error in api._get_axis_size
A user hit [this assertion](https://github.com/google/jax/blob/f6f6b4f04cb1cfb64d531eb41172e1783c57fe47/jax/_src/api.py#L1562). We haven't been able to make a minimal repro yet.

### LINKED ISSUE #10465: Unhelpful error messages related to pmap leading dimension requirements
`pmap` splits work based on the leading dimension of its arguments, which must be equal and cannot be larger than the number of devices. As a JAX n00b, I wasn't initially aware of this, thinking that `pmap` would internally split along the leading dimension. This led me to encounter some unhelpful error messages, one of which was triggered by using keyword arguments.  Repro details below:

```
import jax
import jax.numpy as jnp
import numpy as np
```

Here's a very simple computation that scales a (5,) array by a (1,) array's value:

```
x = np.arange(5)
w = np.array([2.])

def scale(x, w):
  return x * w

scale(x, w)
--> array([0., 2., 4., 6., 8.])
```
Note that I'm doing this on a single CPU:

```
jax.local_device_count() 
--> 1
```

Running this computation using `pmap` is inherently broken, because the first dimension of x is 5, which is greater than the number of devices.

So, we'd expect a helpful error message, but I've found that if arguments are supplied by keyword, we don't get one:

```
jax.pmap(scale)(x=x, w=w)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
jax/_src/api.py in _mapped_axis_size(tree, vals, dims, name, kws)
   1499   try:
-> 1500     size, = mapped_axis_sizes
   1501     return size

ValueError: too many values to unpack (expected 1)

During handling of the above exception, another exception occurred:

UnfilteredStackTrace                      Traceback (most recent call last)
6 frames
<ipython-input-7-997e0e638d4f> in <module>()
----> 1 jax.pmap(scale)(x=x, w=w)

jax/_src/traceback_util.py in reraise_with_filtered_traceback(*args, **kwargs)
    161     try:
--> 162       return fun(*args, **kwargs)
    163     except Exception as e:

jax/_src/api.py in cache_miss(*args, **kwargs)
   2011 
-> 2012     out_tree, out_flat = f_pmapped_(*args, **kwargs)
   2013     out_pytree_def = out_tree()

jax/_src/api.py in pmap_f(*args, **kwargs)
   1884   def pmap_f(*args, **kwargs):
-> 1885     p = _prepare_pmap(
   1886         fun, in_axes, out_axes, static_broadcasted_tuple, donate_tuple,

jax/_src/api.py in _prepare_pmap(fun, in_axes, out_axes, static_broadcasted_tuple, donate_tuple, global_arg_shapes, args, kwargs)
   1828       kws=True))
-> 1829   local_axis_size = _mapped_axis_size(
   1830       in_tree, args, in_axes_flat, "pmap", kws=True)

jax/_src/api.py in _mapped_axis_size(tree, vals, dims, name, kws)
   1512       tree, leaf = treedef_children(tree)
-> 1513       assert treedef_is_leaf(leaf)
   1514     # TODO(mattjj,phawkins): add a way to inspect pytree kind more directly

UnfilteredStackTrace: AssertionError

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

AssertionError                            Traceback (most recent call last)
<ipython-input-7-997e0e638d4f> in <module>()
----> 1 jax.pmap(scale)(x=x, w=w)

AssertionError:
```

### PR DESCRIPTION
**Pull request by jakevdp**
Fixes #8883, #10465 and #10193

### CODE REVIEW

File: jax/_src/api.py [hunk 1]
```diff
@@ -1595,8 +1595,16 @@ def _get_axis_size(name: str, shape: Tuple[int, ...], axis: int):
...
       # message only to be about the positional arguments
+      full_tree = tree
       tree, leaf = treedef_children(tree)
-      assert treedef_is_leaf(leaf)
+
+      # Separately ensure that keyword arguments match in size.
+      if not treedef_is_leaf(leaf):
+        sizes = [x.shape[d] if d is not None else None for x, d in zip(vals, dims)]
+        sizes = tree_unflatten(full_tree, sizes)
+        print(repr(msg.format(f"the tree of axis sizes is:\n{sizes}")))
```
Comment (sharadmv) [inline jax/_src/api.py:1605]: Remove print

File: jax/_src/api.py [hunk 2]
```diff
@@ -1595,8 +1595,16 @@ def _get_axis_size(name: str, shape: Tuple[int, ...], axis: int):
     if kws:
       # if keyword arguments are included in the tree, we make adapt the error
```
Comment (sharadmv) [inline jax/_src/api.py:1596]: Is this comment relevant anymore?

File: jax/_src/api.py [hunk 3]
```diff
@@ -1595,8 +1595,16 @@ def _get_axis_size(name: str, shape: Tuple[int, ...], axis: int):
...
+      full_tree = tree
       tree, leaf = treedef_children(tree)
-      assert treedef_is_leaf(leaf)
+
+      # Separately ensure that keyword arguments match in size.
+      if not treedef_is_leaf(leaf):
+        sizes = [x.shape[d] if d is not None else None for x, d in zip(vals, dims)]
+        sizes = tree_unflatten(full_tree, sizes)
+        print(repr(msg.format(f"the tree of axis sizes is:\n{sizes}")))
+        raise ValueError(msg.format(f"the tree of axis sizes is:\n{sizes}")) from None
```
Comment (sharadmv) [inline jax/_src/api.py:1606]: Should this `raise ... from e` like the error above?

**Review by sharadmv — APPROVED** [review]
This is a nice QOL change, thanks!

File: jax/_src/api.py (hunk 2 above)
Comment (jakevdp) [reply jax/_src/api.py:1596]: Yes, it applies to the following line where `tree` is rewritten with a copy that does not include keyword arguments.

File: jax/_src/api.py (hunk 3 above)
Comment (jakevdp) [reply jax/_src/api.py:1606]: I did this to be consistent with the existing `raise ... from None` used further below.

**Pull request merged**